README.md
QtMiniSEED
==========

The Qt wrapper for libmseed (read miniseed from QIODevice)


CHANGELOG
ChangeLog
=========

vNEXT
-----
* correct code

v0.1.0
------
* first release


src/qtminiseedrecord.h
// Copyright 2013-2015, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#ifndef QTMINISEEDRECORD_H
#define QTMINISEEDRECORD_H


#include "qtminiseedrecorddata.h"
#include <QString>


class QtMiniSeedRecord
{
    friend class QtMiniSeedDecoder;
private:
    explicit QtMiniSeedRecord(MSRecord *record);
public:
    QtMiniSeedRecord();
    QtMiniSeedRecord(const QtMiniSeedRecord &other);
    enum DataType { Unknow, Ascii, Int, Float, Double };

    inline bool isNull() const;

    inline const QString network() const;
    inline const QString station() const;
    inline const QString channel() const;
    inline const QString location() const;

    inline qint32 sequenceNumber() const;
    inline qint64 startTime() const;
    inline qint64 endTime() const;
    inline double sampleRate() const;
    inline qint32 sampleCount() const;
    inline DataType sampleType() const;
    inline qint32 sampleSize() const;

    QByteArray toAscii() const;

    inline qint32 recordSize() const;
    inline const char *recordData() const;
    inline void *datasamples() const;

private:
    QExplicitlySharedDataPointer<QtMiniSeedRecordData> d;
};

bool QtMiniSeedRecord::isNull() const
{
    return (d->msrecord == 0);
}

const QString QtMiniSeedRecord::network() const
{
    return d->msrecord->network;
}

const QString QtMiniSeedRecord::station() const
{
    return d->msrecord->station;
}

const QString QtMiniSeedRecord::channel() const
{
    return d->msrecord->channel;
}

const QString QtMiniSeedRecord::location() const
{
    return d->msrecord->location;
}

qint32 QtMiniSeedRecord::sequenceNumber() const
{
    return d->msrecord->sequence_number;
}

qint64 QtMiniSeedRecord::startTime() const
{
    return d->msrecord->starttime;
}

qint64 QtMiniSeedRecord::endTime() const
{
    return d->msrecord->starttime
           + static_cast<hptime_t>(1000000.0
                                   / d->msrecord->samprate
                                   * d->msrecord->samplecnt);
}

double QtMiniSeedRecord::sampleRate() const
{
    return d->msrecord->samprate;
}

qint32 QtMiniSeedRecord::sampleCount() const
{
    return d->msrecord->samplecnt;
}

QtMiniSeedRecord::DataType QtMiniSeedRecord::sampleType() const
{
    switch (d->msrecord->sampletype) {
    case 'a':
        return Ascii;
    case 'i':
        return Int;
    case 'f':
        return Float;
    case 'd':
        return Double;
    }
    return Unknow;
}

qint32 QtMiniSeedRecord::sampleSize() const
{
    switch (d->msrecord->sampletype) {
    case 'a':
        return 1;
    case 'i':
    case 'f':
        return 4;
    case 'd':
        return 8;
    }
    return 0;
}

qint32 QtMiniSeedRecord::recordSize() const
{
    return d->msrecord->reclen;
}

const char *QtMiniSeedRecord::recordData() const
{
    return d->msrecord->record;
}

void *QtMiniSeedRecord::datasamples() const
{
    return d->msrecord->datasamples;
}


#endif // QTMINISEEDRECORD_H


src/qtminiseeddecoder.h
// Copyright 2013-2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#ifndef QTMINISEEDDECODER_H
#define QTMINISEEDDECODER_H


#include "qtminiseedrecord.h"
#include <QIODevice>


/*!
 * \brief The QgeMiniSeedDecoder provides the functionality to decode the miniseed
 * stream into discret miniseed records. \see QgeMiniSeedRecord
 */
class QtMiniSeedDecoder
{
public:
    enum Error {
        NoError = 0,
        FewDataError,
        FormatError,
        ReadError,
        UnexpectedOffsetError,
        UnexpectedEndError,
        UnknowError = 128
    };

    explicit QtMiniSeedDecoder(qint64 expectedSize = 4096, bool blocked = true);
    virtual ~QtMiniSeedDecoder();

    QtMiniSeedRecord readRecord(QIODevice *device);
    QtMiniSeedRecord readRecord(const char *data, int size, int *position);

    void reset();

    inline int lastError() const;
    QString errorString() const;

    static QtMiniSeedRecord decode(char *data, int size, QString *reason = 0);
    static QtMiniSeedRecord decode(const char *data, int size, QString *reason = 0);
    static QtMiniSeedRecord decode(QIODevice *device, int size, QString *reason = 0);

private:
    void resize();
    QtMiniSeedRecord iteration();

private:
    qint64 m_expectedSize;
    qint64 m_realRecordLen;
    char *m_record;
    Error m_lastError;
    bool m_isBlocked;

    qint64 offset;
    // current read processor state
    qint64 asize;
    qint64 rsize;
    qint8 is_end;
    qint8 bitswp;
    // element type
    // 0: FSDH
    // 1: BLKHDR
    // 2: DATA
    // 3: 1000 blk
    qint8 r_type;
};

/*!
 * \brief QgeMiniSeedDecoder::lastError
 * \return last error of the stream processing
 */
int QtMiniSeedDecoder::lastError() const
{
    return m_lastError;
}


#endif // QTMINISEEDDECODER_H


src/qtminiseedrecord.cpp
// Copyright 2013-2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#include "qtminiseedrecord.h"
#include <QMetaType>


QtMiniSeedRecord::QtMiniSeedRecord(MSRecord *record)
{
    qRegisterMetaType<QtMiniSeedRecord>("QtMiniSEEDRecord");
    d = new QtMiniSeedRecordData(record);
}

QtMiniSeedRecord::QtMiniSeedRecord()
{
    qRegisterMetaType<QtMiniSeedRecord>("QtMiniSEEDRecord");
    d = new QtMiniSeedRecordData();
}

QtMiniSeedRecord::QtMiniSeedRecord(const QtMiniSeedRecord &other)
    : d(other.d)
{
}

QByteArray QtMiniSeedRecord::toAscii() const
{
    if (sampleType() == Ascii) {
        return QByteArray::fromRawData(reinterpret_cast<const char *>(d->msrecord->datasamples),
                                       sampleSize() * sampleCount());
    }

    return QByteArray();
}


src/qtminiseedrecorddata.h
// Copyright 2015, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#ifndef QTMINISEEDRECORDDATA_H
#define QTMINISEEDRECORDDATA_H


#include <libmseed.h>
#include <QSharedData>


class QtMiniSeedRecordData : public QSharedData
{
public:
    QtMiniSeedRecordData(MSRecord *rec = 0)
    {
        msrecord = rec;
    }

    ~QtMiniSeedRecordData()
    {
        msr_free(&msrecord);
    }

public:
    MSRecord *msrecord;
};


#endif // QTMINISEEDRECORDDATA_H


src/qtminiseedrecorddata.cpp
// Copyright 2015, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#include "qtminiseedrecorddata.h"


src/qtminiseeddecoder.cpp
// Copyright 2013-2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#include "qtminiseeddecoder.h"


typedef struct __blkhdr_t {
    quint16 type;
    quint16 offset;
}
blkhdr_t;


QtMiniSeedDecoder::QtMiniSeedDecoder(qint64 expectedSize, bool blocked)
{
    m_expectedSize = expectedSize;
    m_record = new char[expectedSize];
    m_isBlocked = blocked;

    reset();
}

QtMiniSeedDecoder::~QtMiniSeedDecoder()
{
    delete []m_record;
}

QtMiniSeedRecord QtMiniSeedDecoder::readRecord(QIODevice *device)
{
    // algorithm
    try {
        while (device->bytesAvailable()) {
            resize();

            qint64 toRead = rsize - asize;
            if (device->bytesAvailable() < toRead)
                toRead = device->bytesAvailable();

            if (device->read(&m_record[offset + asize], toRead) != toRead)
                throw ReadError;

            asize += toRead;
            if (rsize == asize) {
                asize = 0;
                QtMiniSeedRecord record = iteration();
                if (!record.isNull())
                    return record;
            }
        }

        throw FewDataError;
    }
    catch (Error error) {
        m_lastError = error;
    }

    // error
    return QtMiniSeedRecord();
}

QtMiniSeedRecord QtMiniSeedDecoder::readRecord(const char *data, int size, int *position)
{
    // algorithm
    try {
        while (*position < size) {
            resize();

            qint64 toRead = rsize - asize;
            qint64 avaible = size - *position;
            if (avaible < toRead)
                toRead = avaible;

            qMemCopy(&m_record[offset + asize], &data[*position], toRead);
            *position += toRead;

            asize += toRead;
            if (rsize == asize) {
                asize = 0;
                QtMiniSeedRecord record = iteration();
                if (!record.isNull())
                    return record;
            }
        }

        throw FewDataError;
    }
    catch (Error error) {
        m_lastError = error;
    }

    // error
    return QtMiniSeedRecord();
}

void QtMiniSeedDecoder::reset()
{
    offset = 0;
    rsize = sizeof(fsdh_s);
    asize = 0;
    is_end = 0;
    bitswp = 0;
    r_type = 0;
    m_lastError = NoError;
}

QString QtMiniSeedDecoder::errorString() const
{
    switch (lastError()) {
    case NoError:
        return QObject::tr("No Error", "QgeMiniSeedRecord");
    case FewDataError:
        return QObject::tr("Need more data", "QgeMiniSeedRecord");
    }

    return QString();
}

QtMiniSeedRecord QtMiniSeedDecoder::decode(char *data, int size, QString *reason)
{
    Q_UNUSED(reason);

    MSRecord *msrec = 0;
    msr_unpack(data, size, &msrec, true, 1);
    return QtMiniSeedRecord(msrec);
}

QtMiniSeedRecord QtMiniSeedDecoder::decode(const char *data, int size, QString *reason)
{
    char *tmp = new char[size];
    memcpy(tmp, data, size);
    return decode(tmp, size, reason);
}

QtMiniSeedRecord QtMiniSeedDecoder::decode(QIODevice *device, int size, QString *reason)
{
    Q_UNUSED(reason);

    if (device->bytesAvailable() >= size) {
        MSRecord *msrec = 0;
        char *data = new char[size];
        device->read(data, size);
        msr_unpack(data, size, &msrec, true, 1);
        delete []data;
        return QtMiniSeedRecord(msrec);
    }

    return QtMiniSeedRecord();
}

void QtMiniSeedDecoder::resize()
{
    m_realRecordLen = offset + rsize;
    if ((m_realRecordLen  > m_expectedSize && !is_end) || (m_realRecordLen  != m_expectedSize && is_end)) {
        while (m_expectedSize < m_realRecordLen)
            m_expectedSize *= 2;

        char *record = 0;
        record = new char[m_expectedSize];
        qMemCopy(record, m_record, offset);
        delete [] m_record;
        m_record = record;
    }
}

QtMiniSeedRecord QtMiniSeedDecoder::iteration()
{
    // current ptr for quick access
    char *curptr = &m_record[offset];
    offset += rsize;

    // MiniSEED record complitly readed
    if (is_end == 1) {
        // reset to ready next record
        reset();

        // create the miniseed record
        MSRecord *msrec = 0;
        msr_unpack(m_record, m_realRecordLen, &msrec, true, 1);
        return QtMiniSeedRecord(msrec);
    }
    // FSDH readed
    else if (r_type == 0) {
        // check for correct FSDH
        if (!MS_ISVALIDHEADER(curptr))
            throw FormatError;

        // next: blkhdr
        r_type = 1;
        rsize = sizeof(blkhdr_t);

        // Check byteswap needed by year
        if ((((fsdh_s *)curptr)->start_time.year < 1900) || (((fsdh_s *)curptr)->start_time.year > 2050))
            bitswp = 1;
    }
    // blkhdr readed
    else if (r_type == 1) {
        quint16 hdr_type = ((blkhdr_t *)curptr)->type;
        quint16 hdr_offset = ((blkhdr_t *)curptr)->offset;

        // bitswap if needed
        if (bitswp) {
            ms_gswap2(&hdr_type);
            ms_gswap2(&hdr_offset);
        }

        //
        if (hdr_type == 1000) {
            // start read the 1000blk
            r_type = 3;
            //rsize = hdr_offset - offset;
            rsize = sizeof(blkt_1000_s);
        }
        else if (hdr_offset != 0) {
            // next blocket offset cannot point early than the current offset
            if (hdr_offset < offset)
                throw UnexpectedOffsetError;

            // start read the raw data
            r_type = 2;
            rsize = hdr_offset - offset;
        }
        else {
            throw UnexpectedEndError;
        }
    }
    // data readed
    else if (r_type == 2) {
        r_type = 1;
        rsize = sizeof(blkhdr_t);
    }
    // blk1000 breaded
    else if (r_type == 3) {
        is_end = 1;
        rsize = (1 << ((blkt_1000_s *)curptr)->reclen) - offset;
    }

    return QtMiniSeedRecord();
}
